bin/_update-git

```diff
@@ -12,55 +12,106 @@ exec 2>&1
+fi
+
+current_dir=$(pwd)
+echo "Attempting to update project repo ${PROJECT_NAME} at '${current_dir}' -> '$(readlink -e ${current_dir})'"
 
 if [ -d "./.git" ]; then
+    # validate that the current repository matches the expected repository
+    ORIGIN_REPOSITORY=$(git remote get-url origin)
+    if [ "${ORIGIN_REPOSITORY}" != "${PROJECT_REPOSITORY}" ]; then
+	echo "ERROR: Invalid state!  The origin repository does not match the project repository!"
+	exit 1
+    fi
+
+    current_ref=$(git rev-parse --verify HEAD)
+    expected_ref=$(git rev-parse --verify ${PROJECT_CHECKOUT_TARGET})
+
+    if [ "${current_ref}" != "${expected_ref}" ]; then
+	echo "ERROR: Invalid state!  The current checked out ref (${current_ref}) does not match the checkout target's ref (${PROJECT_CHECKOUT_TARGET} -> ${expected_ref})!"
+	exit 1
+    fi
+
     # fetch updates from remote(s)
-    if ! git remote --verbose update --prune > /dev/null 2>&1; then
-	echo "ERROR: could not git-update"
+    git_fetch_cmd="git fetch --all --prune --prune-tags --tags --verbose"
+    echo "${git_fetch_cmd}"
+    if ! ${git_fetch_cmd} 2>&1; then
+	echo "ERROR: could not git-fetch"
 	exit 1
     fi
 
-    git_status=$(git_get_status)
-    git_local_branch=$(git_get_local_branch "${git_status}")
-    git_tracking=$(git_get_tracking "${git_status}")
+    # store any local changes
+    stash_output=$(git stash)
+    if [ $? != 0 ]; then
+	echo "ERROR: could not git-stash"
+	exit 1
+    fi
 
-    if [ -n "${git_tracking}" -a "${git_local_branch}" != "(detached)" ]; then
-	# store any local changes
-	stash_output=$(git stash)
-	if [ $? != 0 ]; then
-	    echo "ERROR: could not git-stash"
-	    exit 1
-	fi
+    # return to the primary branch so that we can then return to the
+    # checkout target in case it's reference has changed (ie. an
+    # updated tag)
+    echo "Switching to primary branch to quiesce the repository state"
+    git checkout ${PROJECT_PRIMARY_BRANCH}
 
-	# merge any changes from the remote branch
-	if ! git pull --verbose --ff-only 2> /dev/null; then
-	    echo "ERROR: could not git-pull"
-	    exit 1
-	fi
+    # return to the checkout target if it is not the primary branch
+    if [ "${PROJECT_PRIMARY_BRANCH}" != "${PROJECT_CHECKOUT_TARGET}" ]; then
+	echo "Returning to the checkout target"
+	git -c advice.detachedHead=false checkout ${PROJECT_CHECKOUT_TARGET}
+    fi
 
-	if ! echo "${stash_output}" | grep -q "No local changes to save"; then
-            # reapply local changes
-            if ! git stash pop > /dev/null; then
-		echo -e "${stash_output}"
-		echo "ERROR: could not git-stash pop"
+    if [ "${PROJECT_CHECKOUT_MODE}" == "locked" ]; then
+	# return to the previous ref if is supposed to be locked there
+	echo "Returning to the previously checked out ref since I am in locked mode"
+	git checkout ${current_ref}
```

Comment:
OK, now it makes more sense to me. The action depends on the mode, and this is not about what we do with the mode.